src/ProcFdDirMonitor.cc

```diff
@@ -0,0 +1,121 @@
+/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include "ProcFdDirMonitor.h"
+
+#include <dirent.h>
+#include <stdlib.h>
+
+#include "AutoRemoteSyscalls.h"
+#include "RecordSession.h"
+#include "RecordTask.h"
+#include "log.h"
+
+using namespace std;
+
+namespace rr {
+
+ProcFdDirMonitor::ProcFdDirMonitor(Task* t, const string& pathname) {
+  // XXX this makes some assumptions about namespaces... Probably fails
+  // if |t| is not the same pid namespace as rr
+  int ends_with_slash = (pathname.back() == '/');
+  if (pathname.substr(0, 6) == string("/proc/") &&
+      pathname.substr(pathname.size() - 3 - ends_with_slash,
+		      3) == string("/fd")) {
+    string s = pathname.substr(6, pathname.size() - 9 - ends_with_slash);
+    char* end;
+    int tid = strtol(s.c_str(), &end, 10);
+    if (!*end) {
+      Task* target = t->session().find_task(tid);
+      if (target) {
+        tuid = target->tuid();
+      }
+    }
+  }
+}
+
+// returns the number of valid dirent structs left in the buffer
+template <typename D>
+static int filter_dirent_structs(Task* t, uint8_t* buf, size_t size, int root_fd) {
+  int bytes = 0;
+  size_t current_offset = 0;
+  size_t offset_adjustment = 0;
+  while (1) {
+    if (current_offset == size) {
+      break;
+    }
+
+    D* current_struct = reinterpret_cast<D*>(buf + current_offset);
+    auto next_off = current_offset + current_struct->d_reclen;
+
+    char* fname = (char*)current_struct->d_name;
+    char* end;
+    int fd = strtol(fname, &end, 10);
+    if (!*end && fd >= root_fd) {
+      // Skip this entry.
+      memmove(current_struct, buf + next_off, size - next_off);
+      offset_adjustment += next_off;
```

Comment:
Good catch. `offset_adjustment` is left over from before I realized that dirents only have relative offsets anyways and should removed.

And yes, you're right.

I suspect this worked because when Chromium enumerates through the fds only the root_fd exists.  I'll write some more tests.